src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala

```diff
@@ -63,27 +67,48 @@ class ResultsUploader private (commitUuid: String, codacyClient: CodacyClient, b
     codacyClient.sendEndOfResults(commitUuid)
   }
 
-  private def uploadResultsBatch(tool: String, batchSize: Int, results: Set[Result]): Future[Either[String, Unit]] = {
-    if (results.size <= batchSize) {
-      codacyClient.sendRemoteResults(tool, commitUuid, results)
-    } else {
-      val batch: Future[Either[String, Unit]] =
-        codacyClient.sendRemoteResults(tool, commitUuid, results.take(batchSize))
-      val remainingBatches: Future[Either[String, Unit]] =
-        uploadResultsBatch(tool, batchSize, results.drop(batchSize))
-      sequenceUploads(batch, remainingBatches)
-    }
+  private def uploadResultsBatch(tool: String,
+                                 batchSize: Int,
+                                 results: Set[FileResults]): Future[Either[String, Unit]] = {
+    val (_, remainingFileResults, uploadResponses) =
+      results.foldLeft((0, Set.empty[FileResults], Seq.empty[Future[Either[String, Unit]]])) {
+        case ((accumulatedResultsSize, accumulatedFileResults, responses), fileResults)
+            if (accumulatedResultsSize + fileResults.results.size) >= batchSize =>
+          val response = codacyClient.sendRemoteResults(tool, commitUuid, accumulatedFileResults)
+          (fileResults.results.size, Set(fileResults), responses :+ response)
+
+        case ((accumulatedResultsSize, accumulatedFileResults, responses), fileResults) =>
+          (accumulatedResultsSize + fileResults.results.size, accumulatedFileResults + fileResults, responses)
+      }
```

Comment:
If we had a way to split into batches and then map the batches into the upload I believe it would end up being more readable than getting the next batch on a fold.